### DISCUSSION TITLE
Would be nice if the copilot would also help in code refactoring work

### DISCUSSION THREAD

**Discussion by Freestyle00** [community/community#10114] [Copilot Conversations] [answered]
This would increase the effectiveness of the Copilot even more as in only creating code but also deleting it and replacing it.

**Comment by oxysoft** [comment, marked as answer]
There needs to be a feature to make prompt templates, e.g.

```
<ORIGINAL_CODE>
---
[Rewrite the above code to <PROMPT>]
<CARET>
```

So you would select a piece of code, enter your prompt, and we can cycle through ways to rewrite the original code that the model spat out. You can do a LOT with this, like "remove the error checking" and it will work 99% of the time if it's self-documenting and everything. Game-changing, in many ways, I can't imagine what the community would come up with. It can fix exceptions and non-compiling code as well, "Rewrite the code to fix <COMPILE_ERROR>".

Some times it needs to be started with the first couple of words of the original code.

You can probably even direct its attention by summarizing data into a pseudocode context block, like the code in external files, but telling only the symbol names with pseudocode:

```
---
Context:
- Foo.fields: damage, physical, magical
- Foo.methods: GetName(), GetBody(), Postprocess()
---
```

**Reply by ARMMWORXS** [reply]
Thanks. Nicolas.

**Reply by cwilby** [reply]
Reuse existing intellisense functionality.  When you select some text, show a "lightbulb" that includes an action that triggers GitHub to refactor the code.

The actual code refactoring would be a case of "delete what's there and try again CoPilot"

Maybe the code is daft half the time, but I would guess it would be pretty useful most of the time.

**Reply by rvsfara** [reply]
yes allow accept changes

**Reply by andraz** [reply]
This is partially available in brushes as Chunk action.

It works for me 25% of the time to refactor, most of the time it just destroys code, but occasionally it works if you select right amount of lines. (my field is React JSX)

**Reply by hems** [reply]
> This is partially available in brushes as Chunk action.

what you mean "in brushes as chunk action" ?

**Reply by andraz** [reply]
Copilot Labs

**Reply by Anselmoo** [reply]
With the new  [Copilot's Custom-Feature](https://github.com/orgs/community/discussions/13901#discussioncomment-4110167), you can define it as a command; just type in the field `refactor`.

<img width="90" alt="image" src="https://user-images.githubusercontent.com/13209783/211400211-af1a9e26-7f4f-44bf-a21f-c853b3701360.png">

**Reply by sanikolov** [reply]
Is there an equivalent of the `refactor` command in VS2022? I poked around the Copilot extension, there were no brushes or textbox widgets for entering commands.

**Reply by fssrepository** [reply]
To be honest, most of the work what a programmer does on a transformation project is to refactor. If i can say to Copilot, browse up the loaded project and refactor them to microservices and deploy to cloud system within an instant. Create some diagram, what TMforum defines with plantuml, i would retire right away. That task can be highly automated. I'm not looking for brushes, do it on my workspace, in VSCode with git it can't ruin anything and with gitlense i can compare and review step by step, what it has done, before commit.

**Reply by ayim** [reply]
@sanikolov the build in IntelliCode feature helps with specific refactorings (i.e. repeated edits) via the PROSE engine.
The upcoming Copilot Chat for VS offering will provide a prompt based refactoring experience, link here:
https://devblogs.microsoft.com/visualstudio/github-copilot-chat-for-visual-studio-2022/

**Comment by kpripper** [comment]
Copilot labs is not avaliable now.
